File: data/derived_cot/rq3_prompting/results/translation/r1/py_to_cpp/code_output/Calculator.cpp
#include <cctype>
#include <cmath>
#include <optional>
#include <string>
#include <vector>
#include <stdexcept>

class Calculator {
public:
    Calculator() = default;

    std::optional<double> calculate(const std::string& expression) {
        std::vector<double> operand_stack;
        std::vector<char> operator_stack;
        std::string num_buffer;

        for (char ch : expression) {
            if (std::isdigit(ch) || ch == '.') {
                num_buffer += ch;
            } else {
                if (!num_buffer.empty()) {
                    operand_stack.push_back(std::stod(num_buffer));
                    num_buffer.clear();
                }

                if (ch == '+' || ch == '-' || ch == '*' || ch == '/' || ch == '^') {
                    while (!operator_stack.empty() &&
                           operator_stack.back() != '(' &&
                           precedence(operator_stack.back()) >= precedence(ch)) {
                        apply_operator(operand_stack, operator_stack);
                    }
                    operator_stack.push_back(ch);
                } else if (ch == '(') {
                    operator_stack.push_back(ch);
                } else if (ch == ')') {
                    while (!operator_stack.empty() && operator_stack.back() != '(') {
                        apply_operator(operand_stack, operator_stack);
                    }
                    if (!operator_stack.empty()) {
                        operator_stack.pop_back(); // remove '('
                    }
                }
                // ignore any other characters
            }
        }

        if (!num_buffer.empty()) {
            operand_stack.push_back(std::stod(num_buffer));
        }

        while (!operator_stack.empty()) {
            apply_operator(operand_stack, operator_stack);
        }

        if (operand_stack.empty()) {
            return std::nullopt;
        }
        return operand_stack.back();
    }

private:
    int precedence(char op) const {
        switch (op) {
            case '+':
            case '-': return 1;
            case '*':
            case '/': return 2;
            case '^': return 3;
            default:  return 0;
        }
    }

    void apply_operator(std::vector<double>& operand_stack,
                        std::vector<char>& operator_stack) const {
        char op = operator_stack.back();
        operator_stack.pop_back();

        double operand2 = operand_stack.back();
        operand_stack.pop_back();
        double operand1 = operand_stack.back();
        operand_stack.pop_back();

        double result;
        switch (op) {
            case '+': result = operand1 + operand2; break;
            case '-': result = operand1 - operand2; break;
            case '*': result = operand1 * operand2; break;
            case '/':
                if (operand2 == 0.0) {
                    throw std::runtime_error("Division by zero");
                }
                result = operand1 / operand2;
                break;
            case '^':
                if (operand1 < 0 && std::fmod(operand2, 1.0) != 0.0) {
                    throw std::runtime_error("Negative base with fractional exponent");
                }
                result = std::pow(operand1, operand2);
                break;
            default:
                throw std::runtime_error("Unknown operator");
        }
        operand_stack.push_back(result);
    }
};

File: data/derived_cot/rq2_eval/translation/r1/_converted_code/cpp_to_py/Calculator.py
import math

def safe_div(x, y):
    try:
        return x / y
    except ZeroDivisionError:
        if x != x:  # Check if x is NaN
            return float('nan')
        if x == 0:
            return float('nan')
        if (x < 0) != (y < 0):
            return -float('inf')
        else:
            return float('inf')

def safe_pow(x, y):
    try:
        return x ** y
    except ZeroDivisionError:
        if x == 0 and y < 0:
            k = -y
            if k.is_integer():
                if math.copysign(1, x) < 0 and int(k) % 2 == 1:
                    return -float('inf')
                else:
                    return float('inf')
            else:
                return float('inf')
        else:
            raise

class Calculator:
    def __init__(self):
        self.operators = {
            '+': lambda x, y: x + y,
            '-': lambda x, y: x - y,
            '*': lambda x, y: x * y,
            '/': safe_div,
            '^': safe_pow
        }
    
    def precedence(self, op):
        if op in ['+', '-']:
            return 1
        elif op in ['*', '/']:
            return 2
        elif op == '^':
            return 3
        else:
            return 0
    
    def apply_operator(self, operand_stack, operator_stack):
        op_stack_copy = operator_stack[:]
        operand_stack_copy = operand_stack[:]
        op = op_stack_copy.pop()
        operand2 = operand_stack_copy.pop()
        operand1 = operand_stack_copy.pop()
        if op not in self.operators:
            raise ValueError(f"Unknown operator: {op}")
        func = self.operators[op]
        result = func(operand1, operand2)
        operand_stack_copy.append(result)
        return operand_stack_copy, op_stack_copy

    def calculate(self, expression):
        if not expression:
            return 0.0
        operand_stack = []
        operator_stack = []
        num_buffer = ''
        for ch in expression:
            if ch.isdigit() or ch == '.':
                num_buffer += ch
            else:
                if num_buffer:
                    operand_stack.append(float(num_buffer))
                    num_buffer = ''
                if ch in ['+', '-', '*', '/', '^']:
                    while (operator_stack and 
                           operator_stack[-1] != '(' and 
                           self.precedence(operator_stack[-1]) >= self.precedence(ch)):
                        operand_stack, operator_stack = self.apply_operator(operand_stack, operator_stack)
                    operator_stack.append(ch)
                elif ch == '(':
                    operator_stack.append(ch)
                elif ch == ')':
                    while operator_stack and operator_stack[-1] != '(':
                        operand_stack, operator_stack = self.apply_operator(operand_stack, operator_stack)
                    if operator_stack and operator_stack[-1] == '(':
                        operator_stack.pop()
        if num_buffer:
            operand_stack.append(float(num_buffer))
        while operator_stack:
            operand_stack, operator_stack = self.apply_operator(operand_stack, operator_stack)
        return operand_stack[-1] if operand_stack else 0.0